### PR TITLE
retry

### LINKED ISSUE #361: 503 response on docker pull
By to mohlo udělat retry. Je to teda z dockerhubu, ale mohlo by se to stát i pro ECR?

image

### LINKED ISSUE #390: improve ecr retry
- retry longer up to 10-20 mins? 
- retry everything except
```
"class":"InvalidArgumentException","message":"Found 1 error while validating the input provided for the GetAuthorizationToken operation:\n[registryIds][0] must be a string or an object that implements __toString(). Found bool(false)"
```
which happens for misconfigured components

### PR DESCRIPTION
**Pull request by odinuv**
feat: more aggressive retry for pulling images, longer timeouts and unified timeouts across all providers, fixes #390, fixes #361

### CODE REVIEW

**Comment by odinuv** [comment]
| delay | attempt |  
-- | -- | -- | --
  | 100 | 1 |  
  | 200 | 2 |  
  | 400 | 3 |  
  | 800 | 4 |  
  | 1600 | 5 |  
  | 3200 | 6 |  
  | 6400 | 7 |  
  | 12800 | 8 |  
  | 25600 | 9 |  
  | 51200 | 10 |  
  | 60000 | 11 |  
  | 60000 | 12 |  
  | 60000 | 13 |  
  | 60000 | 14 |  
  | 60000 | 15 |  
  | 60000 | 16 |  
  | 60000 | 17 |  
  | 60000 | 18 |  
  | 60000 | 19 |  
  | 60000 | 20 |  
  | 702300 |   |  
total | 11.705 | mins |  
  |   |   |

**Comment by odinuv** [comment]
added `setRetryLimits` to make the tests pass within travis 10minute limit

File: src/Docker/Image/AWSElasticContainerRegistry.php [hunk 1]
```diff
@@ -91,8 +98,8 @@ public function getLogoutParams()
      */
     protected function pullImage()
     {
-        $retryPolicy = new SimpleRetryPolicy(3);
-        $backOffPolicy = new ExponentialBackOffPolicy(10000);
+        $retryPolicy = new SimpleRetryPolicy($this->retryMaxAttempts);
+        $backOffPolicy = new ExponentialBackOffPolicy($this->retryMinInterval, 2, $this->retryMaxInterval);
         $proxy = new RetryProxy($retryPolicy, $backOffPolicy);
```
Comment (ondrejhlavacek) [inline src/Docker/Image/AWSElasticContainerRegistry.php]: celý vytvářený RetryProxy hodit do vlastní metody a do parent classy?

File: src/Docker/Image.php [hunk 2]
```diff
@@ -58,6 +58,21 @@ abstract class Image
      */
     private $imageDigests;
 
+    /**
+     * @var int
+     */
+    protected $retryMinInterval = 100;
```
Comment (ondrejhlavacek) [inline src/Docker/Image.php]: není tohle zbytečně málo? začal bych třeba na vteřině.

File: src/Docker/Image/AWSElasticContainerRegistry.php (hunk 1 above)
Comment (odinuv) [reply src/Docker/Image/AWSElasticContainerRegistry.php]: jo, dobrej napad

File: src/Docker/Image.php (hunk 2 above)
Comment (odinuv) [reply src/Docker/Image.php]: dal jsem tak kompropmis 500ms, timpadem se to na sekundu dostane pri druhym retry (predtim az po 4 retry), cili takto:

  | delay | attempt
-- | -- | --
  | 500 | 1
  | 1000 | 2
  | 2000 | 3
  | 4000 | 4
  | 8000 | 5
  | 16000 | 6
  | 32000 | 7
  | 60000 | 8
  | 60000 | 9
  | 60000 | 10
  | 60000 | 11
  | 60000 | 12
  | 60000 | 13
  | 60000 | 14
  | 60000 | 15
  | 60000 | 16
  | 60000 | 17
  | 60000 | 18
  | 60000 | 19
  | 60000 | 20
  | 843500 |  
total | 14.05833 | mins

**Pull request merged**